### PR TITLE
feat: Improve authentication adapter interface

### LINKED ISSUE #7042: Modernize Auth Adapter
**Is your feature request related to a problem? Please describe.**
Not a problem

**Describe the solution you'd like**
We can introduce webauthn adapter to support modern auth system

**Additional context**
We can set up this with https://simplewebauthn.dev/ 

An additional PR can be done on the Parse-JS SDK to support WebAuthn sign up

### PR DESCRIPTION
**Pull request by dblythy**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description

Continues on from #7079.

Related issue: #7079

### Approach

- Adds ability to throw errors from
- cleans up code
- removes parameter `options` as it can be accessed on the adapter via `this`

### TODOs before merging

- [x] Add tests
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

### CODE REVIEW

**Comment by mtrezza** [comment]
@Moumouls Do you want to review this PR before merging?

**Comment by Moumouls** [comment]
Yes @mtrezza , I'll check the diff between this one and my original PR

**Comment by Moumouls** [comment]
Quite busy currently @mtrezza, @dblythy  but i don't forget this one 👍 
I will also build a custom version and test it on my company repo to check that everything works with advanced usage

**Comment by mtrezza** [comment]
@Moumouls Did you get to try this out?

**Comment by mtrezza** [comment]
A gentle reminder to @Moumouls and @rhuanbarreto that this PR is ready for review. I remember you both have been very outspoken in https://github.com/parse-community/parse-server/pull/7079 about the urgency of getting this feature merged. It's now waiting for >1 month for feedback on the new changes, so if you could take a look and post some feedback that would be great, so we can finally get this merged. Thanks.

File: src/Adapters/Auth/AuthAdapter.js
Comment (Moumouls) [inline src/Adapters/Auth/AuthAdapter.js]: We need to keep `options` to be able to control the auth adapter from the outside. Parse Server always supported this strategy and it works well.

File: src/Adapters/Auth/AuthAdapter.js
Comment (Moumouls) [inline src/Adapters/Auth/AuthAdapter.js]: keep options

File: src/Adapters/Auth/AuthAdapter.js
Comment (Moumouls) [inline src/Adapters/Auth/AuthAdapter.js]: keep options

File: src/Adapters/Auth/AuthAdapter.js
Comment (Moumouls) [inline src/Adapters/Auth/AuthAdapter.js]: keep options

File: src/Adapters/Auth/AuthAdapter.js
Comment (Moumouls) [inline src/Adapters/Auth/AuthAdapter.js]: Keep options
`validateAuthData(authData, options, request) {`

File: src/Adapters/Auth/AuthAdapter.js
Comment (Moumouls) [inline src/Adapters/Auth/AuthAdapter.js]: keep options and request

File: src/Adapters/Auth/index.js
Comment (Moumouls) [inline src/Adapters/Auth/index.js:84]: interesting here you kept options, this is the right thing

File: src/Adapters/Auth/index.js
Comment (Moumouls) [inline src/Adapters/Auth/index.js]: keep options

File: src/Adapters/Auth/index.js
Comment (Moumouls) [inline src/Adapters/Auth/index.js]: keep options

File: src/Adapters/Auth/index.js
Comment (Moumouls) [inline src/Adapters/Auth/index.js]: keep options

File: src/Adapters/Auth/index.js
Comment (Moumouls) [inline src/Adapters/Auth/index.js]: keep options

File: src/Auth.js
Comment (Moumouls) [inline src/Auth.js]: you can early "exit" here with `continue`

File: src/Auth.js
Comment (Moumouls) [inline src/Auth.js]: Bug found here `return acc;`

By chance i found it using this PR in my project. I'm really not a huge fan of the combination of `for` `continue` `mutating` and early return strategies. It could lead to these kinds of critical bugs, hard to catch sometimes because it's related to a functional branch and how many loops `for` will perform.

I strongly suggest reverting to my original code using the `reducePromise` strategy, for safer execution by design using the functional approach that i used. (also on `challenge` part.)

Also we need to write a new test that cover the bug that i found.

The process of the fail:
- Create auth adapter A, auth adapter B
- Create/Setup a user with adapter A
- then update the user with `{authData: { adapterA: null, adapterB: {id: "test"}}}`
- BUG: if you get the user after the update, you should notice that user has no `adapterB` in authData

Note: this bug could be flaky depending of the `provider` name. Since we sort() keys for the `for`. Using provider names like `providerA` and `providerB` should be stable to reproduce this critical bug.

**Review by Moumouls — CHANGES_REQUESTED** [review]
Critical bug found due to removal of the `functional approach`

**Comment by Moumouls** [comment]
So @mtrezza if the options and the functional approach are restored we should be okay (plus the new test). I did not find other issues running this branch on my company project. We are close. Thanks, @dblythy  for the continuation.

File: src/Adapters/Auth/AuthAdapter.js
Comment (dblythy) [reply src/Adapters/Auth/AuthAdapter.js]: I figured options could be removed as they can be accessed via using `this` on the adapter

File: src/Auth.js
Comment (dblythy) [reply src/Auth.js]: Good pickup!

**Comment by mtrezza** [comment]
Let's try to merge this PR by end of next week. There are some other changes that are blocked by this PR, so it would be great if we could accomplish that.

**Comment by Moumouls** [comment]
@dblythy @mtrezza I'll take a look into this today, we are so close :rocket:

Thanks @dblythy for you contributions !

**Review by Moumouls — DISMISSED** [review]
LGTM, tested latest changes on my envs, it seems that we are good to go.

One test to fix,  i think it's related to the revert of options.

File: src/Options/index.js
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* Allow a user to log in even if the 3rd party authentication token that was used to sign in to their account has expired. If this is set to `false`, then the token will be validated every time the user signs in to their account. This refers to the token that is stored in the `_User.authData` field. Defaults to `true`.
```

Requires `npm run definitions`

**Review by mtrezza — COMMENTED** [review]
Is this description correct? If yes, please commit and run `npm run definitions`.

**Comment by mtrezza** [comment]
Do / should we have any docs that come with this PR?

To confirm again, this PR doesn't include any breaking changes? If it does, that's also fine since we can release it with Parse Server 6, but we would need to document them.

File: src/Options/index.js
Comment (dblythy) [reply src/Options/index.js]: @Moumouls could you please review 😊

Comment (Moumouls) [reply src/Options/index.js]: i think your suggestion is okay @mtrezza , more detailed. In the futur we need to switch `default: false`

But we can't currently since it could be a huge breaking change.

**Review by Moumouls — DISMISSED** [review]
@dblythy just use commit suggestion of @mtrezza  and we should be good to go !

File: DEPRECATIONS.md
Comment (mtrezza) [inline DEPRECATIONS.md]: The deprecation descriptions are all a bit unclear. It's unclear whether they describe how it's before or after the deprecation removal. For now let's continue to describe how it will be after the removal, e.g. "Login with expired 3rd party authentication token defaults to `false`." In the future we could improve the table in general.

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Comment by Moumouls** [comment]
@mtrezza I forgot your comment

> To confirm again, this PR doesn't include any breaking changes?

No breaking changes here, Parse is still fully compatible with current and older auth systems. Basically it's just a new auth interface for backend adapters.


BTW, we can consider to release it into Parse V6 as a new major feature, since with this system it will be really easy to develop Webauthn, OTP, TOTP, MFA, and any advanced auth mechanism !

**Comment by mtrezza** [comment]
It's quite a significant change so I would also say to release it with PS6, not as a feature in PS5. In case there are any issues, people can stay on PS5 and we can still maintain the LTS branch separately.

In that sense we got this ready for merge just in time 😉

File: DEPRECATIONS.md
Comment (mtrezza) [inline DEPRECATIONS.md]:
```suggestion
| DEPPS8 | Login with expired 3rd party authentication token defaults to `false` | [#7079](https://github.com/parse-community/parse-server/pull/7079)   | 5.3.0 (2022)                    | 7.0.0 (2024)                    | deprecated            | -     |
```

Comment (mtrezza) [reply DEPRECATIONS.md]: fixed

**Review by mtrezza — APPROVED** [review]
Looks good, is everyone fine with merging this?

And could anyone come up with a better PR title? The PR title will be used as the changelog entry, so it'd be a good opportunity to advertise this feature a bit more.

**Comment by mtrezza** [comment]
Could anyone come up with a better PR title? The PR title will be used as the changelog entry, so it'd be a good opportunity to advertise this feature a bit more.

**Comment by Moumouls** [comment]
@mtrezza i can suggest "feat: enhanced auth adapter interface" 

Btw, let's merge this one !

**Comment by mtrezza** [comment]
Could we find a more descriptive changelog entry? A title that gives a better idea about the new capabilities, so developers can better understand what benefit this change has, and they start exploring the feature, for example MFA. 4 words seems a pretty modest description for a feature that required so much work 😉

How about:

> feat: Improve authentication adapter interface to support multi-factor authentication (MFA), authentication challenges, and provide a more powerful interface for writing custom authentication adapters.

**Comment by Moumouls** [comment]
you description is nice @mtrezza let's go for it 🚀

**Comment by mtrezza** [comment]
Great, we have a few other PRs to merge before this; I estimate we'll merge this on Monday.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-alpha.2](https://github.com/parse-community/parse-server/releases/tag/6.0.0-alpha.2)

**Comment by mtrezza** [comment]
It's merged! Thanks @Moumouls for starting this and @dblythy for finishing this, and everyone else between.

**Comment by dblythy** [comment]
Awesome work @Moumouls - keen to see the cool new auth adapters that are made thanks to your work! Would be good to try ship some of these new ones as part of Parse Server 6

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/6.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0](https://github.com/parse-community/parse-server/releases/tag/6.0.0)

**Pull request merged**
